src/s1ap/handlers/handover_required.c

```diff
@@ -0,0 +1,160 @@
+                    ho_required_ies.data[i].val.enb_ue_s1ap_id;
+
+        }
+            break;
+        case S1AP_IE_HANDOVER_TYPE:
+        {
+            log_msg(LOG_INFO, "handover required S1AP_IE_HANDOVER_TYPE.\n");
+
+            ho_required.msg_data.handover_required_Q_msg_m.handoverType =
+                    ho_required_ies.data[i].val.handoverType;
+        }
+            break;
+        case S1AP_IE_CAUSE:
+        {
+            log_msg(LOG_INFO, "handover required S1AP_IE_CAUSE.\n");
+
+            memcpy(&ho_required.msg_data.handover_required_Q_msg_m.cause,
+                    &ho_required_ies.data[i].val.cause,
+                    sizeof(struct s1apCause));
+        }
+            break;
+        case S1AP_IE_TARGET_ID:
+        {
+            log_msg(LOG_INFO, "handover required S1AP_IE_TARGET_ID.\n");
+
+            memcpy(&ho_required.msg_data.handover_required_Q_msg_m.target_id,
+                    &ho_required_ies.data[i].val.target_id,
+                    sizeof(struct targetId));
+
+            char *enb_id_buf = ho_required_ies.data[i].val.target_id.global_enb_id.macro_enb_id;
+
+            enb_id =
+                (enb_id_buf[0] << 12) +
+                (enb_id_buf[1] << 4) + ((enb_id_buf[2] & 0xf0) >> 4);
+
+            log_msg(LOG_INFO, "enbId %d\n", enb_id);
+
+        }
+            break;
+        case S1AP_IE_SOURCE_TOTARGET_TRANSPARENTCONTAINER:
+        {
+            log_msg(LOG_INFO,
+                    "handover required S1AP_IE_SOURCE_TOTARGET_TRANSPARENTCONTAINER.\n");
+
+            ho_required.msg_data.handover_required_Q_msg_m.srcToTargetTranspContainer.count =
+                    ho_required_ies.data[i].val.srcToTargetTranspContainer.size;
+
+            memcpy(
+                    ho_required.msg_data.handover_required_Q_msg_m.srcToTargetTranspContainer.buffer,
+                    ho_required_ies.data[i].val.srcToTargetTranspContainer.buffer_p,
+                    ho_required_ies.data[i].val.srcToTargetTranspContainer.size);
+        }
+            break;
+        default:
+            log_msg(LOG_WARNING, "Unhandled IE");
+        }
+    }
+
+    if (enb_id == 0)
+    {
+        log_msg(LOG_ERROR, "Failed to decode target enb id %d", enb_id);
+        return E_FAIL;
+    }
+
+    uint32_t cbIndex = findControlBlockWithEnbId(enb_id);
+    if (cbIndex == INVALID_CB_INDEX)
+    {
+        log_msg(LOG_ERROR, "Failed to find cb for enb id %d", enb_id);
+        return E_FAIL;
+    }
+    ho_required.msg_data.handover_required_Q_msg_m.target_enb_fd = cbIndex;
+
+    ho_required.msg_type = handover_required;
+    ho_required.destInstAddr = htonl(mmeAppInstanceNum_c);
+    ho_required.srcInstAddr = htonl(s1apAppInstanceNum_c);
+
+    int i = send_tipc_message(ipc_S1ap_Hndl, mmeAppInstanceNum_c,
+            (char*) &ho_required,
+            S1_READ_MSG_BUF_SIZE);
+
+    if (i < 0)
+    {
+        log_msg(LOG_ERROR, "Error To write in s1_handover_required_handler\n");
+    }
+    //TODO: free IEs
```

Comment:
Can you please explain what's missing ? Please open a user story and put it in handover bucket.

Reply:
Changed to free the memory allocated during decode.